stdr.go

```diff
@@ -105,264 +104,59 @@ type StdLogger interface {
-	File string `json:"file"`
-	Line int    `json:"line"`
-}
-
-func (l logger) caller() callerID {
-	// +1 for this frame, +1 for Info/Error.
-	_, file, line, ok := runtime.Caller(l.depth + 2)
-	if !ok {
-		return callerID{"<unknown>", 0}
-	}
-	return callerID{filepath.Base(file), line}
-}
-
-func (l *logger) Init(info logr.RuntimeInfo) {
-	l.depth += info.CallDepth
-}
-
 func (l logger) Enabled(level int) bool {
 	return globalVerbosity >= level
 }
 
 func (l logger) Info(level int, msg string, kvList ...interface{}) {
-	args := make([]interface{}, 0, 64) // using a constant here impacts perf
-	if l.logCaller == All || l.logCaller == Info {
-		args = append(args, "caller", l.caller())
+	prefix, args := l.FormatInfo(level, msg, kvList)
+	if prefix != "" {
+		args = prefix + ": " + args
 	}
-	args = append(args, "level", level, "msg", msg)
-	args = append(args, l.values...)
-	args = append(args, kvList...)
-	argsStr := flatten(args...)
-	l.output(l.depth+1, fmt.Sprintln(l.prefix, argsStr))
+	l.output(args)
 }
 
 func (l logger) Error(err error, msg string, kvList ...interface{}) {
-	args := make([]interface{}, 0, 64) // using a constant here impacts perf
-	if l.logCaller == All || l.logCaller == Error {
-		args = append(args, "caller", l.caller())
-	}
-	args = append(args, "msg", msg)
-	var loggableErr interface{}
-	if err != nil {
-		loggableErr = err.Error()
-	}
-	args = append(args, "error", loggableErr)
-	args = append(args, l.values...)
-	args = append(args, kvList...)
-	argsStr := flatten(args...)
-	l.output(l.depth+1, fmt.Sprintln(l.prefix, argsStr))
-}
-
-func (l logger) output(calldepth int, s string) {
-	depth := calldepth + 2 // offset for this adapter
-
-	// ignore errors - what can we really do about them?
-	if l.std != nil {
-		_ = l.std.Output(depth, s)
-	} else {
-		_ = log.Output(depth, s)
+	prefix, args := l.FormatError(err, msg, kvList)
+	if prefix != "" {
+		args = prefix + ": " + args
 	}
+	l.output(args)
 }
 
-// WithName returns a new logr.Logger with the specified name appended.  stdr
-// uses '/' characters to separate name elements.  Callers should not pass '/'
-// in the provided name string, but this library does not actually enforce that.
 func (l logger) WithName(name string) logr.LogSink {
-	if len(l.prefix) > 0 {
-		l.prefix += "/"
-	}
-	l.prefix += name
+	l.Formatter.AddName(name)
 	return &l
-
 }
 
-// WithValues returns a new logr.Logger with the specified key-and-values
-// saved.
 func (l logger) WithValues(kvList ...interface{}) logr.LogSink {
-	// Three slice args forces a copy.
-	n := len(l.values)
-	l.values = append(l.values[:n:n], kvList...)
+	l.Formatter.AddValues(kvList)
 	return &l
-
 }
 
 func (l logger) WithCallDepth(depth int) logr.LogSink {
-	l.depth += depth
+	l.Formatter.AddCallDepth(depth)
 	return &l
 }
 
+func (l logger) output(s string) {
+	depth := l.Formatter.GetDepth() + 2 // offset for this adapter
+
+	// ignore errors - what can we really do about them?
+	if l.std != nil {
```

Comment:
Would it make sense to avoid this if check and then perhaps the entire `output` helper function by ensuring in the constructor that `l.std` is always non-nil? If `NewWithOptions` is called with a nil interface, the constructor could create one which calls `log.Output`.

If `l.std` continues to be nil in some cases, then `GetUnderlying` might need a comment about that.

Reply:
Good idea.  Will add a commit